### DISCUSSION TITLE
Sobre programar em alguma distribuição Linux

### DISCUSSION THREAD

**Discussion by ludani** [frontendbr/forum#359] [Discussão]
# **O que vcs acham sobre programar em linux ?**
   Sou fã da distribuição em linux... uso a mais de 4 anos , mas vamos ser sincero... usar terminal para programar ou até mesmo instalar pacotes de programação no linux. Requer além de paciência um estudo mais profundo  dos pacotes para fazer qualquer alteração no sistema. 
Isso atrapalha ou ajudaria o Dev ?

**Comment by karlprieb** [comment]
Uso Linux em casa e Mac no trabalho, o terminal sempre me ajudou e facilitou a vida. Instalar um package pelo terminal para mim é muito mais rápido do que baixar um instalador, clicar next, next, next, etc, etc.

Manter esses pacotes atualizados também é muito mais fácil com um gerenciador de pacotes do que instalar binarios na mão :)

Não vejo nenhum contra no uso de gerenciador de pacotes. Quais estão sendo tuas dificuldades?

**Comment by icaromh** [comment]
Nossa, tu é a primeira pessoa que eu vi na vida a dizer que usar o terminal requer paciência

**Comment by giioohbernini** [comment]
Paciência é preciso para programar no Windows, hehe. Mas concordo que uma
hora ou outra você terá problemas em alguma coisa, que de certa for vai te
dar uma dor de cabeça para resolver, como algum driver que não esta
funcionando corretamente, algum software não tem para Linux.. Mas na minha
opinião compensa muito!

Em ter, 8 de nov de 2016 14:54, Ícaro Heimig notifications@github.com
escreveu:

**Comment by woliveiras** [comment]
Oloko!

Até no Mac eu prefiro utilizar o Terminal pra quase tudo.

Só não uso de fato para programar, pois ainda sou mais produtivo no SublimeText.

Concordo com o pessoal... Entrar no site, baixar um instalador pra depois instalar alguma coisa eu acho bem menos produtivo. Fora que pode vir com um monte de pacotes que não serão úteis e vão ficar ali sujando seu sistema. Ficar limpando SO é muito mais chato! kkk

**Comment by fabiomirandaa** [comment]
Cara, eu uso ElementaryOS, baseado em Ubuntu LTS, e é sensacional a beleza do sistema e a funcionalidade do terminal. Com o terminal você instala as coisas bem rápido, você tem uma comunidade bem grande de soluções já prontas pro Linux. 

Eu lembro em 2014, por exemplo, que era uma bosta usar node e o mongodb no Windows, e por isso fui pro linux. Desde então nunca mais saí do Linux.

**Reply by maxxcleiton** [reply]
ElementaryOS baseado no Ubuntu?
N é baseado no Arch?
Ou era diferente em 2016

**Reply by fdaciuk** [reply]
@maxxcleiton talvez vc esteja confundindo com EndeavourOS. Elementary sempre foi Ubuntu based =)

**Comment by ranosi** [comment]
Cara eu acho que o maior problema com relação ao Linux é o primeiro contato
com ele. Quando você tá iniciando e vê alguem mais avançado digitando 300
comandos te assusta, mas depois que tu vai ganhando intimidade a coisa flui
tranquilamente.
Eu to voltando a usar agora, to curtindo. Única coisa que realmente preciso
do Windows hoje em dia é o Photoshop, fora isso to de boas com o penguin.

Em 8 de novembro de 2016 16:25, Fábio Miranda notifications@github.com
escreveu:

> Cara, eu uso ElementaryOS, baseado em Ubuntu LTS, e é sensacional a beleza
> do sistema e a funcionalidade do terminal. Com o terminal você instala as
> coisas bem rápido, você tem uma comunidade bem grande de soluções já
> prontas pro Linux.
> 
> Eu lembro em 2014, por exemplo, que era uma bosta usar node e o mongodb no
> Windows, e por isso fui pro linux. Desde então nunca mais saí do Linux.
> 
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> https://github.com/frontendbr/forum/issues/359#issuecomment-259217649,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/AV1hh8mlJ0dEDSRLZwFsBYwdwmBK3eldks5q8L6PgaJpZM4Ksjwx
> .

## 

Grande abraço,

[image: --]

Rafael Nogueira
[image: https://]about.me/rafaelnogueira
https://about.me/rafaelnogueira?promo=email_sig&utm_source=email_sig&utm_medium=email_sig&utm_campaign=external_links

**Comment by jonasporto** [comment]
Na minha família até minha sogra usa linux (=, sério mesmo!

Me sentia super produtivo com o terminal basicão, o bash mesmo.. Mas agora com ZSH + Guake + TMUX, minha nossa, não quero outra coisa.

**Comment by herberthenrique** [comment]
Eu já ouvi um monte de "dev" reclamando que digitar comando dá mais trabalho.
Enquanto isso eu crio scripts pra automatizar toda minha vida :)

Terminal é incrivel, depois que você começa a usar.. Não tem volta.

**Comment by ranosi** [comment]
@herberthenrique conte mais sobre esses scripts.

**Comment by eduardojmatos** [comment]
@ludani só mudei o título pra galera entender melhor seu ponto, blz?

**Comment by woliveiras** [comment]
@ranosi da uma olhada: https://github.com/woliveiras/dotfiles

**Comment by ranosi** [comment]
@woliveiras valeu cara. Muito bom isso.
E isso pode ser criado pra instalação de qualquer programa no linux né? Depois vou buscar umas informações sobre a criação, achei muito fera cara.

**Comment by karlprieb** [comment]
@ranosi Depende do programa, a maioria usa um file/folder de configuração na home do user. O que facilita syncar configs com estes "dotfiles"

**Comment by herberthenrique** [comment]
@ranosi Eraexatamente disso que o @woliveiras mostrou.  hahahaha

**Comment by gabrieldarezzo** [comment]
Edit: Relatos de um usuário de Windows

A cada atalho/comando no terminal q você aprende é um click a menos. (e segundos a mais pra ti utilizar da forma que bem entender)

Vamos utilizar o npm como exemplo...

Antes dele como se adicionava um Plugin-Jquery (Full-Calendar ) em um projeto?
Você acessava o site...
Baixava o .js, instalava, selecionava a pasta, extraia

Hoje você só da um (ABRIR TERMINAL) -> cd pastaDoProjeto
`npm install fullcalendar` e flw,vlw. todos os arquivos prontinhos lá pra você. 

Outra coisa que reparei ultimamente foi no Sublime por exemplo.
Ele tem alta integração com o sistema de versionamento Git.
Altos atalhos que economizam oq eu fazia antigamente com o SVN via mouse.

Quanto menos tempo gasto em tarefa mecânica, mais tempo para as legais xD

Acredito que algum usuário do VIM poderia dar um exemplo (pratico) de como voar com atalhos para tarefas especificas.

A parte legal tmb da linha de comando é q é possível ir automatizando alguns processos feitos.

Resumo do que falei:
Quanto menos se usa o mouse maior a velocidade.

**Comment by felipefialho** [comment]
Usei Windows por muitos e muitos anos, e só agora mudei pro Mac.

Mesmo no Windows usava MUITO o terminal, mas ainda usava muito o mouse e acesso as pastas via interface.

Agora, no Mac, estou usando o Terminal pra quase tudo, mal uso o mouse. Acho muito mais rapido e pratico.

**Comment by nicklima** [comment]
Uso o Windows para desenvolver no trabalho e sinto uma maior lentidão ao rodar tarefas do GULP nele do que no Linux. 

Não tenho usado o Linux no momento mas já estou preparando minha máquina (em casa) pois pretendo desenvolver no Pinguim quando for trabalhar nos meus projetos pessoais e nos Freelas.

Só de imaginar que basta executar um comando via terminal para instalar programas, executar tarefas e até preparar o ambiente de desenvolvimento me deixa com vontade de por logo esse projeto em prática.

**Comment by RatoX** [comment]
@gabrieldarezzo 

Sou usuário vim já tem um tempinho uso bastante sistemas POSIX e acho que posso falar sobre precisão e velocidade.

Eu uso bastante o velho modal ( H, J, K e L ) e todos os meus key-bindings são baseado neles. Uso junto com o tmux no terminal onde eu tenho controle total sobre os terminais e o que eles devem fazer. 

Um exemplo prático do meu dia-a-dia:
- Tenho um projeto com 5 ambientes diferentes, cada um com seus atalhos e seus ambientes virtuais ( Python )

Começo criando uma sessão no tmux para o ambiente que vou trabalhar:

`tmux new-session -s 'frontend'`

Ai eu abro meu VIM com todos os meus arquivos editados do meu ultimo WIP ( geralmente tenho meus atalhos já configurados para esses comandos ):

`git reset --soft HEAD~1`
`vim -O $(git diff --name-only --cached)`

Então eu coloco os testes para rodar em modo **watch** abring uma nova aba no tmux **`Ctrl-b c`** e deixo rodando. Meu tmux tá no silent mode então quando um test quebra ele coloca um "!" no nome da aba que tá com problemas. Quando eu preciso percorrer o staktrace com código eu faço um merge dos paineis e coloco lado-a-lado **`Ctrl-b j _target_`** e acompanho o staktrace com o yank mode do tmux.

Geralmente algum outro problema aparece uma outra parte do sistema. Crio uma nova sessão no tmux `:new` e geralmente trabalho até eu precisar voltar pra a outra sessão **`Ctrl-b (`** ela vai estar lá perfeitamente rodando tudo o que eu deixei aberto, até o cursor fica no lugar.

Isso tudo sem nunca tirar a mão do teclado e nem tocando no mouse.

P.S.: Eu também mapeio o CASPLOCK para ser o meu ESC ( sério isso vai tirar todas as dores do seu pulso ) e mapeio o tmux para ter uma movimentação estilo VIM entre os splits panels. Ctrl+h Ctrl+j Ctrl+k ( os usuários de emacs odeiam isso HAHAHAHAH )

**Comment by gabrieldarezzo** [comment]
@RatoX  é disso que eu tava falando!!!!
Quando eu crescer quero ficar igual a você kkkk

@ludani 

> Isso atrapalha ou ajudaria o Dev ?

Cara ainda acho que todo atalho disse exige dedicação e estudo como você já falou... 
O Linux possibilita uma infinidades de atalhos e automatizações que após passar a curva de aprendizado, mas após passada essa árdua tarefa acho que ninguém bate..

**Comment by RatoX** [comment]
@gabrieldarezzo 

Nem cara, eu só sou preguiçoso mesmo.

**Comment by ranosi** [comment]
Caras vou aproveitar essa discussão pra pedir uma ajuda. Preciso do
VirtualBox ou VMWare pra rodar um Windows aqui, e não emula de jeito
nenhum. Vi que tem um problema ai do VM com Ubuntu 16.04. Alguém conseguiu
resolver?

2016-11-11 15:11 GMT-02:00 Rodrigo Vieira de Alencar <
notifications@github.com>:

> @gabrieldarezzo https://github.com/gabrieldarezzo
> 
> Nem cara, eu só sou preguiçoso mesmo.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> https://github.com/frontendbr/forum/issues/359#issuecomment-260005713,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/AV1hh5mkAi4-utn_XHtu5L2YNe7mg3Opks5q9KG5gaJpZM4Ksjwx
> .

## 

Grande abraço,

[image: --]

Rafael Nogueira
[image: https://]about.me/rafaelnogueira
https://about.me/rafaelnogueira?promo=email_sig&utm_source=email_sig&utm_medium=email_sig&utm_campaign=external_links

**Comment by fdaciuk** [comment]
@RatoX já uso o vim a algum tempinho. 

O @danjesus me apresentou o tmux a alguns meses, mas só comecei a usar agora (desde ontem na real xD), pois conheci o [byobu](http://byobu.co/). 

E cara, na boa: não sei como eu consegui ficar tanto tempo trabalhando só no terminal puro! tmux é sensacional! O Byobu ainda facilita bastante pra quem quer começar com tmux, pois tem alguns atalhos mais simples já pré-configurados..

Mas já estou vendo que é só o pontapé inicial pra usar o tmux de vez xD

Uma outra ferramenta bastante interssante, relacionada ao tmux que eu achei hoje, foi [esse cara](https://github.com/tmux-plugins/tmux-resurrect), pra que você possa salvar suas sessões ainda após reiniciar o PC =)

**Comment by fdaciuk** [comment]
@ranosi, abre uma nova issue, só pra não desfocar do assunto aqui =)

**Comment by RatoX** [comment]
@fdaciuk Cara que foda!! Bem vindo ao tmux, eu tenho um conf file dele nos meus repo tá até bem legal.

Uma dica: instale o vimium (Chrome) ou vimperator (Firefox) para navegar nos sitios dentro da rede de computadores! Vei sua produtividade vai lá pra cima.

**Comment by RatoX** [comment]
@fdaciuk não conhecia o byobu dei uma olhada aqui e ele me pareceu meio raio laser para matar barata mas acredito que deve ajudar no inicio sim.

Como sou vimzeiro minhas key-bindings são um pouco mais mnemônicas tipo `Ctrl-b s` para **split**

**Comment by fdaciuk** [comment]
> @fdaciuk Cara que foda!! Bem vindo ao tmux, eu tenho um conf file dele nos meus repo tá até bem legal.

Com certeza vou olhar =)

> Uma dica: instale o vimium (Chrome) ou vimperator (Firefox) 

Eu tentei usar, mas no navegador não consegui me acostumar.. pra mim atrapalhou um pouco haha

> não conhecia o byobu dei uma olhada aqui e ele me pareceu meio raio laser para matar barata mas acredito que deve ajudar no inicio sim

Pois é.. pelo que vi, ele é só uma interface para o tmux mesmo (dá pra usar o screen no lugar do tmux também), mas agiliza bastante pra começar.. os atalhos do tmux são muito pra minha cabeça por enquanto.. estava acostumado com os atalhos do Terminator pra fazer split, então a curva do Terminator para o Byobu foi menor que ir direto para o tmux.. mas dá pra usar comandos do tmux normal com o byobu, por isso escolhi ele pra começar e ir acostumando com o tmux xD

**Comment by tcelestino** [comment]
Sempre trabalhei em ambientes Windows, até comprar meu primeiro Mac e desde de então nunca mais mexi (deve ter mudado bastante). 

Minha experiência com Linux de inicio foi bem ruim. Comecei a mexer com o Kurumin (sim! hahaha) e era muito ruim não ter a facilidade de instalação do Windows (era heavy user da MS). Passei pelo o Fedora e até tentei usar o Debian, não deu. Com o surgimento do Ubuntu, voltei para o Linux (mas ainda trabalhava em ambiente Windows), de lá pra cá comecei a curtir mais, mexer com terminal realmente era chato, mas se acostumando, descobrindo atalhos e fazendo a produtividade aumentar ainda mais. 

Hoje vejo que ambientes Linux são mais flexiveis e as distribuições são bem mais faceis de usar, sem contar que é bem mais fácil configurar ambientes ja que a grande maioria dos projetos derivam da mesma arquitetura.

**Comment by felipeuntill** [comment]
Sou usuário de Mac, Windows e Linux e sou proficiente nas 3 plataformas. Meu sonho é um dia conseguir substituir um atom da vida pelo VIM como meus colegas descolados fazem(Sério)...

**Comment by expalmer** [comment]
O que faz o tmux ? quais as vantagens senhores ?

**Comment by fdaciuk** [comment]
@expalmer eu vi nele algumas vantagens bem interessantes:

- Criação de novas janelas (como se fossem abas, mas dentro do tmux);
- Split vertical e horizontal, sem limites de quantidade;
- Sessões separadas para configuração de páginas / splits específicos.

Olhando para esses pontos, pode parecer comum, já que no próprio terminal você pode fazer isso (em alguns terminais, como o Terminator, por exemplo).

Mas a grande vantagem é que ele faz tudo isso dentro da sessão do tmux. Isso te permite, por exemplo, fechar o terminal e manter todas as configurações exatamente igual estavam antes de fechar.

Imagina se o seu terminal dá crash, e você tem várias abas, splits, etc., configurados? Se você trabalha bastante com terminal, vai entender o que estou falando xD

Quando você abrir novamente o terminal, e iniciar o tmux, tudo vai estar como você deixou, inclusive conexões SSH xD

E usando o [`tmux-ressurect`](https://github.com/tmux-plugins/tmux-resurrect) você ainda consegue guardar essas configurações para usar ainda que reinicie seu PC :D

**Comment by danjesus** [comment]
@fdaciuk eu uso tmux a tanto tempo que nem sei mais como é viver sem ele, seja no mac ou linux.

Uma grande vantagem que vejo também é a questão de não usar o mouse o que nos deixa muito mais velozes no dia-a-dia.

**Comment by fdaciuk** [comment]
Pois é @danjesus! Eu conheci pq vc me apresentou xD

**Comment by expalmer** [comment]
Valeu @fdaciuk, vou começar a usar também.

**Comment by fdaciuk** [comment]
@expalmer o byobu é uma boa pedida pra começar.. os atalhos são mais simples, e vc ainda pode usar os atalhos do tmux normalmente, pra ir acostumando com o tempo =)

**Comment by expalmer** [comment]
@fdaciuk Fenomenal, não deixarei de usar mais \o/

**Comment by RatoX** [comment]
@expalmer e @fdaciuk a ideia é se livrar de vez do mouse, ele atrasa nossa vida.
